src/backend/cpu/CMakeLists.txt

```diff
@@ -1,247 +1,273 @@
-    )
-
-LIST(SORT cpp_sources)
-
-source_group(api\\cpp\\Sources FILES ${cpp_sources})
-
-# OS Definitions
-IF(UNIX)
-    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fPIC")
-ELSE(${UNIX}) #Windows
-    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
-ENDIF()
-
-INCLUDE("${CMAKE_CURRENT_SOURCE_DIR}/kernel/sort_by_key/CMakeLists.txt")
-
-IF(DEFINED BLAS_SYM_FILE)
-  ADD_LIBRARY(afcpu_static STATIC
-              ${cpu_headers}
-              ${cpu_sources}
-              ${backend_headers}
-              ${backend_sources}
-              ${SORT_BY_KEY_OBJECTS})
-
-  ADD_LIBRARY(afcpu SHARED
-              ${c_headers}
-              ${c_sources}
-              ${cpp_sources})
-
-  IF(FORGE_FOUND AND NOT USE_SYSTEM_FORGE)
-    ADD_DEPENDENCIES(afcpu_static forge)
-  ENDIF()
-
-  IF(APPLE)
-    SET_TARGET_PROPERTIES(afcpu_static
-        PROPERTIES LINK_FLAGS -Wl,-exported_symbols_list,${BLAS_SYM_FILE})
-    TARGET_LINK_LIBRARIES(afcpu PUBLIC $<TARGET_FILE:afcpu_static>)
-  ELSE(APPLE)
-    add_custom_command(OUTPUT ${PROJECT_BINARY_DIR}/afcpu_static.renamed
-      BYPRODUCTS ${PROJECT_BINARY_DIR}/afcpu_static.renamed
-      COMMAND objcopy --redefine-syms ${BLAS_SYM_FILE} $<TARGET_FILE:afcpu_static> ${PROJECT_BINARY_DIR}/afcpu_static.renamed
-      DEPENDS $<TARGET_FILE:afcpu_static>)
-      TARGET_LINK_LIBRARIES(afcpu PUBLIC ${PROJECT_BINARY_DIR}/afcpu_static.renamed)
-  ENDIF(APPLE)
-
-ELSE(DEFINED BLAS_SYM_FILE)
-
-ADD_LIBRARY(afcpu SHARED
-            ${cpu_headers}
-            ${cpu_sources}
-            ${backend_headers}
-            ${backend_sources}
-            ${c_headers}
-            ${c_sources}
-            ${cpp_sources}
-            ${SORT_BY_KEY_OBJECTS})
-
-ENDIF(DEFINED BLAS_SYM_FILE)
-
-TARGET_LINK_LIBRARIES(afcpu
-                            PRIVATE ${CBLAS_LIBRARIES}
-                            PRIVATE ${FFTW_LIBRARIES}
-                            PRIVATE ${FreeImage_LIBS}
-                     )
-
-IF(LAPACK_FOUND)
-   TARGET_LINK_LIBRARIES(afcpu  PRIVATE ${LAPACK_LIBRARIES})
-ENDIF()
-
-LIST(LENGTH GRAPHICS_DEPENDENCIES GRAPHICS_DEPENDENCIES_LEN)
-IF(${GRAPHICS_DEPENDENCIES_LEN} GREATER 0)
-    ADD_DEPENDENCIES(afcpu ${GRAPHICS_DEPENDENCIES})
-ENDIF(${GRAPHICS_DEPENDENCIES_LEN} GREATER 0)
-
-IF(FORGE_FOUND)
-   TARGET_LINK_LIBRARIES(afcpu PRIVATE ${GRAPHICS_LIBRARIES})
-ENDIF()
-
-SET_TARGET_PROPERTIES(afcpu PROPERTIES
-                      VERSION "${AF_VERSION}"
-                      SOVERSION "${AF_VERSION_MAJOR}")
-
-INSTALL(TARGETS afcpu EXPORT CPU DESTINATION "${AF_INSTALL_LIB_DIR}"
-        COMPONENT libraries)
-
-IF(APPLE)
-    INSTALL(SCRIPT "${PROJECT_SOURCE_DIR}/CMakeModules/osx_install/InstallTool.cmake")
-ENDIF(APPLE)
-
-export(TARGETS afcpu FILE ArrayFireCPU.cmake)
-INSTALL(EXPORT CPU DESTINATION "${AF_INSTALL_CMAKE_DIR}"
-        COMPONENT cmake
-        FILE ArrayFireCPU.cmake)
+add_library(afcpu_static OBJECT "")
+
+# CPU backend source files
+target_sources(afcpu_static
+  PRIVATE
+    Array.cpp
```

Comment:
I have been experimenting with both schemes but I think this approach is better. This way the header and source are in one location and it is easier change both if necessary.